ai2thor/server.py

```diff
@@ -22,6 +23,263 @@ def default(self, obj):
         return super(NumpyAwareEncoder, self).default(obj)
 
 
+class LazyInstanceDetections2D(Mapping):
+    def __init__(self, instance_masks):
+
+        self._detections2d = {}
+        self.instance_masks = instance_masks
+        self._loaded = False
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.instance_masks == other.instance_masks
+        else:
+            return False
+
+    def mask_bounding_box(self, mask):
+        rows = np.any(mask, axis=1)
+        cols = np.any(mask, axis=0)
+        rw = np.where(rows)
+        if len(rw[0]) == 0:
+            return None
+
+        rmin, rmax = rw[0][[0, -1]]
+        cmin, cmax = np.where(cols)[0][[0, -1]]
+
+        return cmin, rmin, cmax, rmax
+
+    def __contains__(self, key):
+        return key in self.instance_masks
+
+    def __getitem__(self, key):
+        if key in self._detections2d:
+            return self._detections2d[key]
+
+        mask = self.instance_masks[key]
+        self._detections2d[key] = self.mask_bounding_box(mask)
+
+        return self._detections2d[key]
+
+    def __len__(self):
+        return len(self.instance_masks)
+
+    def __iter__(self):
+        for k in self.instance_masks.keys():
+            self.__getitem__(k)
+
+        return iter(self._detections2d)
+
+
+class LazyClassDetections2D(LazyInstanceDetections2D):
+
```

Comment:
LGTM.com notes that this class doesn't override `__eq__` which I guess is a potential problem since it adds a new `_loaded` parameter. That said, I don't think this is a problem as you check that everything is subclassed properly.

Reply:
I went ahead and implemented the `__eq__` method to silence the lgtm alert.